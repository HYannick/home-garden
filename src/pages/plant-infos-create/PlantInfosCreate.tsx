/** @jsx jsx */
import React, { Fragment } from 'react';
import { jsx } from '@emotion/core';
import { useTranslation } from 'react-i18next';
import { withRouter } from 'react-router';
import ActionBar from '../../layout/ActionBar';
import { SideLayer } from '../../layout/SideLayer';
import PlantInfosForm from '../../components/plant-infos-form/PlantInfosForm';
import { LevelData, PlantInfosProps } from './PlantInfosCreate.types';

const PlantInfosCreate: React.FC = ({ history, location }: any) => {
  const { t } = useTranslation();
  const { state } = location;
  const initialValues: PlantInfosProps = {
    name: '',
    latin_name: '',
    family: '',
    description: '',
    exposure_level: LevelData.LOW,
    exposure_description: '',
    temperature_level: LevelData.LOW,
    temperature_description: '',
    watering_level: LevelData.LOW,
    watering_description: '',
    soil_type: '',
    tips: '',
    ...state && state.previousValues
  };
  const goToCreateForm = (values: PlantInfosProps) => history.push({pathname: '/create', state: {plantInfos: values}});

  return (
    <Fragment>
      <ActionBar title={t('plant_create_title')} path="/search"/>
      <PlantInfosForm onSubmit={goToCreateForm} initialValues={initialValues} submitLabel={t('plant_infos_form.button.submit')}/>
      <SideLayer fullHeight/>
    </Fragment>
  );
};

export default withRouter(PlantInfosCreate);
