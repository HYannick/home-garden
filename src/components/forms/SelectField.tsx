/** @jsx jsx */
import React, { Fragment } from 'react';
import styled from '@emotion/styled';
import { jsx } from '@emotion/core';
import { LevelData } from '../../pages/plant-infos-create/PlantInfosCreate.types';
import { Label } from './Label';
import ArrowSelect from '../../core/svg/ArrowSelect';

interface FieldProps {
  field?: any,
  form?: any,
  placeholder?: string,
  label?: string,
  type: string,
  selectors: { label: string, value: LevelData }[],
  defaultValue: LevelData
}

const SelectWrapper = styled('div')`
  border: 0.3rem solid ${({ theme }) => theme.palette.grey.light};
  border-radius: 1.5rem;
  max-width: 15rem;
  width: 100%;
  font-size: 1.6rem;
  display: flex;
  align-items: center;
  select {
    width: 100%;
    border: none;
    box-shadow: none;
    background: none;
    outline: ${({ theme }) => theme.palette.primary.light};
    padding: 1rem 2rem;
    font-weight: bold;
    text-align: center;
    font-size: 1.4rem;
    -webkit-appearance: none;
  }
`;

const SelectIcon = styled('div')`
  margin-top: 0.5rem;
  margin-right: 0.8rem;
  svg {
    width: 2rem;
    height: 2rem;
    path {
      fill: ${({ theme }) => theme.palette.grey.dark};
    }
  }
`;

const SelectField: React.FC<FieldProps> = ({ field, label, defaultValue, selectors }) => (
  <Fragment>
    <Label htmlFor={field.name}>{label}</Label>
    <SelectWrapper>
      <select
        id={field.name} name={field.name} defaultValue={defaultValue} value={field.value}
        onChange={field.onChange}>
        {
          selectors.map(({ label, value }) => (
            <option key={value} value={value}>{label}</option>
          ))
        }
      </select>
      <SelectIcon>
        <ArrowSelect/>
      </SelectIcon>
    </SelectWrapper>
  </Fragment>
);

export default SelectField;
