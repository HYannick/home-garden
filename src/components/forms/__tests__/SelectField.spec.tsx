import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup, fireEvent } from '@testing-library/react';
import { renderWithTheme } from '../../../theme-wrapper';
import SelectField from '../SelectField';
import { LevelData } from '../../../pages/plant-infos-create/PlantInfosCreate.types';

afterEach(cleanup);
describe('SelectField', () => {
  const props = {
    field: {
      name: 'puke',
      onChange: jest.fn(),
    },
    defaultValue: LevelData.LOW,
    selectors: [
      {
        label: 'low',
        value: LevelData.LOW,
      },
      {
        label: 'medium',
        value: LevelData.MEDIUM,
      },
      {
        label: 'high',
        value: LevelData.HIGH,
      },
    ],
    type: 'text',
    label: 'Should I puke?',
  };
  it('should render properly', () => {
    const { container } = renderWithTheme(
      <SelectField {...props}/>,
    );
    expect(container).toMatchSnapshot();
  });
  it('should render the right label', () => {
    const { getByText } = renderWithTheme(
      <SelectField {...props}/>,
    );
    expect(getByText(/^Should I puke/)).toHaveTextContent('Should I puke');
  });
  it('should trigger onChange', () => {
    const { getByLabelText } = renderWithTheme(
      <SelectField {...props}/>,
    );
    const input = getByLabelText('Should I puke?');
    fireEvent.change(input, { target: { value: LevelData.MEDIUM } });
    expect(props.field.onChange).toHaveBeenCalledTimes(1);
  });
});
