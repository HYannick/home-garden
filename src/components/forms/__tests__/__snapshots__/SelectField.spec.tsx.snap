// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField should render properly 1`] = `
<div>
  <label
    class="css-npa6bc"
    for="puke"
  >
    Should I puke?
  </label>
  <div
    class="css-19or460"
  >
    <select
      id="puke"
      name="puke"
    >
      <option
        selected=""
        value="LOW"
      >
        low
      </option>
      <option
        value="MEDIUM"
      >
        medium
      </option>
      <option
        value="HIGH"
      >
        high
      </option>
    </select>
    <div
      class="css-gw542d"
    >
      <svg
        viewBox="0 0 10 4"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5.625 3.5a1 1 0 0 1-1.25 0L2.226 1.781A1 1 0 0 1 2.851 0h4.3a1 1 0 0 1 .625 1.781z"
        />
      </svg>
    </div>
  </div>
</div>
`;
